### PR TITLE
Add support for other file extensions in rewrite rules

### PR DESCRIPTION
**Pull request by eternoendless**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Read below
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | I don't know
| Related PRs       | ~
| How to test?      | Read below
| Possible impacts? | ~

## What this is about

There is a known issue where even if your choose to use a format different than jpg, thumbnails will be generated with the `.jpg` extension. This PR doesn't fix that. It updates rewrite rules so that this will be possible in the future (when the thumbnails will use the appropriate file extension.

This PR also simplifies the rewrite rules, because there's a limit of 9 capturing groups and I needed one more to capture the file's extension. To be safe, I added tests. It was not easy because not only `Tools::generateHtaccess()` is not unit testable, but it also does much more than the scope of this change required. So I copied the rules into the unit test. You could argue that this is cheating because it's not actually testing the real code – I did the best that I could with the time I had.

Interesting: I noticed that the original rewrite rules for categories accept URLs which are probably not wanted, like `img/c/-.jpg` or `img/c/1234567-soMe.Th1n*g_cool22-99.jpg`. This should probably be subject to a fix in another PR.

## How to test

1. Install the shop. Alternatively, go to Advanced parameters > performance, then toggle [this switch](https://user-images.githubusercontent.com/1009343/183975960-ded468dd-050a-4600-b013-30dd27d45f17.png) on or off, and save to trigger the regeneration of the .htaccess file at the root of the project. If you do that, switch the toggle again to leave it as it was (or else it could produce side effects that you might blame on my PR).
2. Product and category images on FO should show up normally (as opposed to be 404 not found).

### CODE REVIEW

**Comment by Hlavtox** [comment]
Product Original image ✅ 
https://domain.cz/53832/waytoplay-silnice-highway-24-kusu.jpg

Product Thumbnails  ✅ 
https://domain.cz/53832-medium_default/waytoplay-silnice-highway-24-kusu.jpg
https://domain.cz/53832-medium_default/waytoplay-silnice-highway-24-kusu.webp

Category thumbnails - 403 forbidden 🔴 
https://domain.cz/c/62-category_default/hracky-pro-holky.jpg

**Comment by eternoendless** [comment]
Sorry, I had forgotten to push [6c05518](https://github.com/PrestaShop/PrestaShop/pull/29304/commits/6c05518b807d014ee8edb811041e3de232520c28) last night, which fixes that exact problem. Can you please try again?

**Comment by Hlavtox** [comment]
Fixed now, all good! 🚀

**Review by Hlavtox — APPROVED** [review]

**Comment by matks** [comment]
Re-running failing CI

**Review by matks — APPROVED** [review]

File: tests/Unit/Classes/ToolsTest.php [hunk 1]
```diff
@@ -826,4 +826,270 @@ public function providerArrayInsertElementAfterKey(): iterable
...
     }
+
+    /**
+     * This test verifies the rewrite rules to be written in the .htaccess file against the expected url segments
+     *
+     * @see Tools::generateHtaccess
+     *
+     * @dataProvider provideHtaccessRules
+     */
+    public function testHtaccessRewriteRules(string $rule, string $replacement, array $testCases)
```
Comment (Amoifr) [inline tests/Unit/Classes/ToolsTest.php:837]: do we add return types in legacy part ?

File: tests/Unit/Classes/ToolsTest.php [hunk 2]
```diff
@@ -826,4 +826,270 @@ public function providerArrayInsertElementAfterKey(): iterable
...
+            }
+
+            if ($case['shouldMatch']) {
+                $result = preg_replace($rule, $replacement, $case['uri']);
+                $this->assertSame($case['rewritten'], $result);
+            }
+        }
+    }
+
+    public function provideHtaccessRules()
```
Comment (Amoifr) [inline tests/Unit/Classes/ToolsTest.php:854]: same comment

**Review by Amoifr — COMMENTED** [review]
I'm still impressed by men that can write regexp like that ;-)

**Review by nicosomb — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @

Thank you for your PR, I tested it and it works, as we can see here :

[Untitled_ Oct 24, 2022 3_36 PM.webm](https://user-images.githubusercontent.com/100699445/197540105-b496676b-a90a-4cbf-8d30-507bfa2e0cf3.webm)

Thanks

**Comment by jolelievre** [comment]
Thanks @eternoendless and @AureRita

**Pull request merged**